File: packages/java/src/main/java/com/yourdudeken/mpesa/services/B2CService.java
package com.yourdudeken.mpesa.services;

import com.yourdudeken.mpesa.config.MpesaConfig;
import com.yourdudeken.mpesa.http.HttpClient;
import com.yourdudeken.mpesa.auth.AuthService;
import java.util.HashMap;
import java.util.Map;

public class B2CService {
    private HttpClient httpClient;
    private AuthService auth;
    private MpesaConfig config;
    private String baseUrl;

    public B2CService(HttpClient httpClient, AuthService auth, MpesaConfig config) {
        this.httpClient = httpClient;
        this.auth = auth;
        this.config = config;
        this.baseUrl = httpClient.getBaseUrl();
    }

    public String send(String phonenumber, String commandId, int amount, String remarks,
                      String resultUrl, String timeoutUrl, String shortCodeType) {
        String url = baseUrl + "/mpesa/b2c/v1/paymentrequest";
        Map<String, Object> body = new HashMap<>();
        body.put("InitiatorName", config.getInitiatorName());
        body.put("SecurityCredential", generateSecurityCredential());
        body.put("CommandID", commandId);
        body.put("Amount", amount);
        body.put("PartyA", config.getB2cShortcode());
        body.put("PartyB", phoneValidator(phonenumber));
        body.put("Remarks", remarks);
        body.put("Occassion", "");
        body.put("ResultURL", resultUrl != null ? resultUrl : config.getB2cResultUrl());
        body.put("QueueTimeOutURL", timeoutUrl != null ? timeoutUrl : config.getB2cTimeoutUrl());

        String token = auth.getAccessToken(shortCodeType);
        return httpClient.post(url, body, token);
    }

    public String validated(String phonenumber, String commandId, int amount, String remarks,
                          String idNumber, String resultUrl, String timeoutUrl, String shortCodeType) {
        String url = baseUrl + "/mpesa/b2cvalidate/v2/paymentrequest";
        Map<String, Object> body = new HashMap<>();
        body.put("InitiatorName", config.getInitiatorName());
        body.put("SecurityCredential", generateSecurityCredential());
        body.put("CommandID", commandId);
        body.put("Amount", amount);
        body.put("PartyA", config.getB2cShortcode());
        body.put("PartyB", phoneValidator(phonenumber));
        body.put("Remarks", remarks);
        body.put("Occassion", "");
        body.put("OriginatorConversationID", getFormattedTimestamp());
        body.put("IDType", "01");
        body.put("IDNumber", idNumber);
        body.put("ResultURL", resultUrl != null ? resultUrl : config.getB2cResultUrl());
        body.put("QueueTimeOutURL", timeoutUrl != null ? timeoutUrl : config.getB2cTimeoutUrl());

        String token = auth.getAccessToken(shortCodeType);
        return httpClient.post(url, body, token);
    }

    private String phoneValidator(String phoneNo) {
        if (phoneNo.startsWith("+")) phoneNo = phoneNo.substring(1);
        if (phoneNo.startsWith("0")) phoneNo = "254" + phoneNo.substring(1);
        else if (phoneNo.startsWith("7")) phoneNo = "254" + phoneNo;
        return phoneNo;
    }

    private String getFormattedTimestamp() {
        java.time.LocalDateTime now = java.time.LocalDateTime.now();
        return String.format("%04d%02d%02d%02d%02d%02d",
            now.getYear(), now.getMonthValue(), now.getDayOfMonth(),
            now.getHour(), now.getMinute(), now.getSecond());
    }

    private String generateSecurityCredential() {
        return "";
    }
}

File: packages/dotnet/src/Yourdudeken.Mpesa/Services/B2CService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Yourdudeken.Mpesa.Config;
using Yourdudeken.Mpesa.Http;
using Yourdudeken.Mpesa.Auth;

namespace Yourdudeken.Mpesa.Services
{
    public class B2CService
    {
        private readonly HttpClientWrapper _httpClient;
        private readonly AuthService _auth;
        private readonly MpesaConfig _config;
        private readonly string _baseUrl;

        public B2CService(HttpClientWrapper httpClient, AuthService auth, MpesaConfig config)
        {
            _httpClient = httpClient;
            _auth = auth;
            _config = config;
            _baseUrl = httpClient.GetBaseUrl();
        }

        public async Task<string> Send(string phonenumber, string commandId, int amount, string remarks,
            string resultUrl = null, string timeoutUrl = null, string shortCodeType = "B2C")
        {
            var url = $"{_baseUrl}/mpesa/b2c/v1/paymentrequest";
            var body = new Dictionary<string, object>
            {
                { "InitiatorName", _config.InitiatorName },
                { "SecurityCredential", GenerateSecurityCredential() },
                { "CommandID", commandId },
                { "Amount", amount },
                { "PartyA", _config.B2cShortcode },
                { "PartyB", PhoneValidator(phonenumber) },
                { "Remarks", remarks },
                { "Occassion", "" },
                { "ResultURL", resultUrl ?? _config.B2cResultUrl },
                { "QueueTimeOutURL", timeoutUrl ?? _config.B2cTimeoutUrl }
            };

            var token = await _auth.GetAccessToken(shortCodeType);
            return await _httpClient.Post(url, body, token);
        }

        private string PhoneValidator(string phoneNo)
        {
            if (phoneNo.StartsWith("+")) phoneNo = phoneNo.Substring(1);
            if (phoneNo.StartsWith("0")) phoneNo = "254" + phoneNo.Substring(1);
            else if (phoneNo.StartsWith("7")) phoneNo = "254" + phoneNo;
            return phoneNo;
        }

        private string GenerateSecurityCredential() => "";
    }
}
